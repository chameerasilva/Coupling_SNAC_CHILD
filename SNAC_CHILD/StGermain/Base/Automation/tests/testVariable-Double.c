/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003, Victorian Partnership for Advanced Computing (VPAC) Ltd, 110 Victoria Street, Melbourne, 3053, Australia.
**
** Authors:
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	Patrick D. Sunter, Software Engineer, VPAC. (pds@vpac.org)
**	Luke J. Hodkinson, Computational Engineer, VPAC. (lhodkins@vpac.org)
**	Siew-Ching Tan, Software Engineer, VPAC. (siew@vpac.org)
**	Alan H. Lo, Computational Engineer, VPAC. (alan@vpac.org)
**	Raquibul Hassan, Computational Engineer, VPAC. (raq@vpac.org)
**
**  This library is free software; you can redistribute it and/or
**  modify it under the terms of the GNU Lesser General Public
**  License as published by the Free Software Foundation; either
**  version 2.1 of the License, or (at your option) any later version.
**
**  This library is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
**  Lesser General Public License for more details.
**
**  You should have received a copy of the GNU Lesser General Public
**  License along with this library; if not, write to the Free Software
**  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
**
** Role:
** 	Tests the Double macros of Variable. Will test a single scalar, vector and each scalar created from the vector.
**
** $Id: testVariable-Double.c 2136 2004-09-30 02:47:13Z PatrickSunter $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#include <mpi.h>
#include "Base/Foundation/Foundation.h"
#include "Base/IO/IO.h"
#include "Base/Container/Container.h"
#include "Base/Automation/Automation.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>


int main( int argc, char *argv[] ) {
	int		rank;
	int		procCount;
	int		procToWatch;
	Stream*		stream;
	
	/* Initialise MPI, get world info */
	MPI_Init( &argc, &argv );
	MPI_Comm_size( MPI_COMM_WORLD, &procCount );
	MPI_Comm_rank( MPI_COMM_WORLD, &rank );
	
	BaseFoundation_Init( &argc, &argv );
	BaseIO_Init( &argc, &argv );
	BaseContainer_Init( &argc, &argv );
	BaseAutomation_Init( &argc, &argv );

	stream = Journal_Register( Info_Type, "testVariable" );
	
	if( argc >= 2 ) {
		procToWatch = atoi( argv[1] );
	}
	else {
		procToWatch = 0;
	}
	if( rank == procToWatch ) {
		typedef double Triple[3];

		double* array;
		Triple* structArray;
		Index length = 10;

		double testValues[] = { 123456789.0, 0.987654321 };
		Index testValueCount = 2;
		Index test_I;
		double testValue;

		Variable* var;
		Variable* vec;
		Variable* vecVar[3];

		Variable_Register* reg;

		int i, j;

		array = Memory_Alloc_Array( double, length, "test" );
		structArray = Memory_Alloc_Array( Triple, length, "test" );

		reg = Variable_Register_New();

		var = Variable_NewScalar(
			"Scalar",
			Variable_DataType_Double,
			&length,
			(void**)&array,
			reg );

		vec = Variable_NewVector(
			"Three",
			Variable_DataType_Double,
			3,
			&length,
			(void**)&structArray,
			reg,
			"a",
			"b",
			"c" );
		vecVar[0] = Variable_Register_GetByName( reg, "a" );
		vecVar[1] = Variable_Register_GetByName( reg, "b" );
		vecVar[2] = Variable_Register_GetByName( reg, "c" );

		Variable_Register_BuildAll( reg );

		Print( reg, stream );

		for ( test_I = 0; test_I < testValueCount; ++test_I ) {	

			testValue = testValues[test_I];

			Journal_Printf( stream, "Setting value %lf\n", (double)(double)testValue );

			for ( i = 0; i < length; ++i ) {
				Variable_SetValueDouble( var, i, testValue );

				Variable_SetValueAtDouble( vec, i, 0, testValue );
				Variable_SetValueAtDouble( vec, i, 1, testValue );
				Variable_SetValueAtDouble( vec, i, 2, testValue );
			}

			Journal_Printf( stream, "~~~Scalar~~~\n" );
			for ( i = 0; i < length; ++i ) {
				Journal_Printf( stream, "i=%d: %d %d %d %d %d \n",
						i,
						Variable_GetValueDouble( var, i ) == (double)(double)testValue,
						Variable_GetValueDoubleAsChar( var, i ) == (char)(double)testValue,
						Variable_GetValueDoubleAsShort( var, i ) == (short)(double)testValue,
						Variable_GetValueDoubleAsInt( var, i ) == (int)(double)testValue,
						Variable_GetValueDoubleAsFloat( var, i ) == (float)(double)testValue );
			}

			Journal_Printf( stream, "~~~Vector~~~\n" );
			for ( i = 0; i < length; ++i ) {
				Journal_Printf( stream, "a, i=%d: %d %d %d %d %d ",
						i,
						Variable_GetValueAtDouble( vec, i, 0 ) == (double)(double)testValue,
						Variable_GetValueAtDoubleAsChar( vec, i, 0 ) == (char)(double)testValue,
						Variable_GetValueAtDoubleAsShort( vec, i, 0 ) == (short)(double)testValue,
						Variable_GetValueAtDoubleAsInt( vec, i, 0 ) == (int)(double)testValue,
						Variable_GetValueAtDoubleAsFloat( vec, i, 0 ) == (float)(double)testValue );
				Journal_Printf( stream, "ptr test %s\n",
						( Variable_GetPtrAtDouble( vec, i, 0 ) == &structArray[i][0] ) ? "OK" : "FAILED" );

				Journal_Printf( stream, "b, i=%d: %d %d %d %d %d ",
						i,
						Variable_GetValueAtDouble( vec, i, 1 ) == (double)(double)testValue,
						Variable_GetValueAtDoubleAsChar( vec, i, 1 ) == (char)(double)testValue,
						Variable_GetValueAtDoubleAsShort( vec, i, 1 ) == (short)(double)testValue,
						Variable_GetValueAtDoubleAsInt( vec, i, 1 ) == (int)(double)testValue,
						Variable_GetValueAtDoubleAsFloat( vec, i, 1 ) == (float)(double)testValue );
				Journal_Printf( stream, "ptr test %s\n",
						( Variable_GetPtrAtDouble( vec, i, 1 ) == &structArray[i][1] ) ? "OK" : "FAILED" );

				Journal_Printf( stream, "c, i=%d: %d %d %d %d %d ",
						i,
						Variable_GetValueAtDouble( vec, i, 2 ) == (double)(double)testValue,
						Variable_GetValueAtDoubleAsChar( vec, i, 2 ) == (char)(double)testValue,
						Variable_GetValueAtDoubleAsShort( vec, i, 2 ) == (short)(double)testValue,
						Variable_GetValueAtDoubleAsInt( vec, i, 2 ) == (int)(double)testValue,
						Variable_GetValueAtDoubleAsFloat( vec, i, 2 ) == (float)(double)testValue );
				Journal_Printf( stream, "ptr test %s\n",
						( Variable_GetPtrAtDouble( vec, i, 2 ) == &structArray[i][2] ) ? "OK" : "FAILED" );

			}

			Journal_Printf( stream, "~~~Vector: Sub-Variable~~~\n" );
			for ( i = 0; i < length; ++i ) {
				for ( j = 0; j < 3; ++j ) {
					Journal_Printf( stream, "j=%d, i=%d: %s\n",
						j,
						i,
						( _Variable_GetStructPtr( vecVar[j], i ) == &structArray[i][j] ) ? "OK" : "FAIL" );
				}
			}

			
		}
	
	}
	
	BaseAutomation_Finalise();
	BaseContainer_Finalise();
	BaseIO_Finalise();
	BaseFoundation_Finalise();

	
	/* Close off MPI */
	MPI_Finalize();
	
	return 0; /* success */
}
