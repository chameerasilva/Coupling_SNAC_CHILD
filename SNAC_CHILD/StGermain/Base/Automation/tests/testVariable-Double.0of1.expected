Variable_Register (ptr): 0x805b7a0
	count: 5
	_size: 8
	_delta: 8
	_variable (ptr): 0x805bf48
Stg_Component (ptr): 0x805bf78
Stg_Object (ptr): 0x805bf78
	Stg_Class (ptr): 0x805bf78
		sizeOfSelf: 152
		_deleteSelf: Yes
		type: Variable
		_delete (func ptr): 0xb7fb44a0
		_print (func ptr): 0xb7fb4538
		_copy (func ptr): 0xb7fb4921
	name: Scalar
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7fb3986
	_construct (func ptr): 0xb7fb55c2
	_build (func ptr): 0xb7fb50a2
	_initialise (func ptr): 0xb7fb55b8
	_execute (func ptr): 0xb7fb55bd
	_destroy (func ptr): 0xb7fb5b1a
	isConstructed: True
	isBuilt: True
	isInitialised: False
	hasExecuted: False
	isDestroyed: False
	constructor function name: Scalar-Construct
	build function name: Scalar-Build
	initialise function name: Scalar-Initialise
	execute function name: Scalar-Execute
	destroy function name: Scalar-Destroy
Variable (ptr): (0x805bf78)
	offsetCount: 1
	offsets (ptr): 0x805bcd0
		offsets[0]: 0
	dataTypes (ptr): 0x805bc48
		dataTypes[0]: 5
	dataTypeCounts (ptr): 0x805c430
		dataTypeCounts[0]: 1
	dataSizes (ptr): 0x805d128
		dataSizes[0]: 8
	components (ptr): 0x805c440
		components[0] (ptr): (nil)
	structSize: 8
	structSizePtr (ptr): (nil)
	arrayPtr (ptr): 0x805bde8
	arrayPtrPtr (ptr): 0xbfcdcdec
	arraySize: 10
	arraySizePtr (ptr): 0xbfcdcde4
	allocateSelf = False
Stg_Component (ptr): 0x805c450
Stg_Object (ptr): 0x805c450
	Stg_Class (ptr): 0x805c450
		sizeOfSelf: 152
		_deleteSelf: Yes
		type: Variable
		_delete (func ptr): 0xb7fb44a0
		_print (func ptr): 0xb7fb4538
		_copy (func ptr): 0xb7fb4921
	name: Three
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7fb3986
	_construct (func ptr): 0xb7fb55c2
	_build (func ptr): 0xb7fb50a2
	_initialise (func ptr): 0xb7fb55b8
	_execute (func ptr): 0xb7fb55bd
	_destroy (func ptr): 0xb7fb5b1a
	isConstructed: True
	isBuilt: True
	isInitialised: False
	hasExecuted: False
	isDestroyed: False
	constructor function name: Three-Construct
	build function name: Three-Build
	initialise function name: Three-Initialise
	execute function name: Three-Execute
	destroy function name: Three-Destroy
Variable (ptr): (0x805c450)
	offsetCount: 1
	offsets (ptr): 0x805c740
		offsets[0]: 0
	dataTypes (ptr): 0x805c750
		dataTypes[0]: 5
	dataTypeCounts (ptr): 0x805c760
		dataTypeCounts[0]: 3
	dataSizes (ptr): 0x805d138
		dataSizes[0]: 24
	components (ptr): 0x805c770
		components[0] (ptr): 0x805c788
		components[0]: a
	structSize: 24
	structSizePtr (ptr): (nil)
	arrayPtr (ptr): 0x805be48
	arrayPtrPtr (ptr): 0xbfcdcde8
	arraySize: 10
	arraySizePtr (ptr): 0xbfcdcde4
	allocateSelf = False
Stg_Component (ptr): 0x805c788
Stg_Object (ptr): 0x805c788
	Stg_Class (ptr): 0x805c788
		sizeOfSelf: 152
		_deleteSelf: Yes
		type: Variable
		_delete (func ptr): 0xb7fb44a0
		_print (func ptr): 0xb7fb4538
		_copy (func ptr): 0xb7fb4921
	name: a
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7fb3986
	_construct (func ptr): 0xb7fb55c2
	_build (func ptr): 0xb7fb50a2
	_initialise (func ptr): 0xb7fb55b8
	_execute (func ptr): 0xb7fb55bd
	_destroy (func ptr): 0xb7fb5b1a
	isConstructed: True
	isBuilt: True
	isInitialised: False
	hasExecuted: False
	isDestroyed: False
	constructor function name: a-Construct
	build function name: a-Build
	initialise function name: a-Initialise
	execute function name: a-Execute
	destroy function name: a-Destroy
Variable (ptr): (0x805c788)
	offsetCount: 1
	offsets (ptr): 0x805ca70
		offsets[0]: 0
	dataTypes (ptr): 0x805ca80
		dataTypes[0]: 5
	dataTypeCounts (ptr): 0x805ca90
		dataTypeCounts[0]: 1
	dataSizes (ptr): 0x805d148
		dataSizes[0]: 8
	components (ptr): 0x805caa0
		components[0] (ptr): (nil)
	structSize: 24
	structSizePtr (ptr): (nil)
	arrayPtr (ptr): 0x805be48
	arrayPtrPtr (ptr): 0xbfcdcde8
	arraySize: 10
	arraySizePtr (ptr): 0xbfcdcde4
	allocateSelf = False
Stg_Component (ptr): 0x805cab0
Stg_Object (ptr): 0x805cab0
	Stg_Class (ptr): 0x805cab0
		sizeOfSelf: 152
		_deleteSelf: Yes
		type: Variable
		_delete (func ptr): 0xb7fb44a0
		_print (func ptr): 0xb7fb4538
		_copy (func ptr): 0xb7fb4921
	name: b
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7fb3986
	_construct (func ptr): 0xb7fb55c2
	_build (func ptr): 0xb7fb50a2
	_initialise (func ptr): 0xb7fb55b8
	_execute (func ptr): 0xb7fb55bd
	_destroy (func ptr): 0xb7fb5b1a
	isConstructed: True
	isBuilt: True
	isInitialised: False
	hasExecuted: False
	isDestroyed: False
	constructor function name: b-Construct
	build function name: b-Build
	initialise function name: b-Initialise
	execute function name: b-Execute
	destroy function name: b-Destroy
Variable (ptr): (0x805cab0)
	offsetCount: 1
	offsets (ptr): 0x805cd98
		offsets[0]: 8
	dataTypes (ptr): 0x805cda8
		dataTypes[0]: 5
	dataTypeCounts (ptr): 0x805cdb8
		dataTypeCounts[0]: 1
	dataSizes (ptr): 0x805d158
		dataSizes[0]: 8
	components (ptr): 0x805cdc8
		components[0] (ptr): (nil)
	structSize: 24
	structSizePtr (ptr): (nil)
	arrayPtr (ptr): 0x805be48
	arrayPtrPtr (ptr): 0xbfcdcde8
	arraySize: 10
	arraySizePtr (ptr): 0xbfcdcde4
	allocateSelf = False
Stg_Component (ptr): 0x805cdd8
Stg_Object (ptr): 0x805cdd8
	Stg_Class (ptr): 0x805cdd8
		sizeOfSelf: 152
		_deleteSelf: Yes
		type: Variable
		_delete (func ptr): 0xb7fb44a0
		_print (func ptr): 0xb7fb4538
		_copy (func ptr): 0xb7fb4921
	name: c
	nameAllocationType: NON_GLOBAL
	_defaultConstructor (func ptr): 0xb7fb3986
	_construct (func ptr): 0xb7fb55c2
	_build (func ptr): 0xb7fb50a2
	_initialise (func ptr): 0xb7fb55b8
	_execute (func ptr): 0xb7fb55bd
	_destroy (func ptr): 0xb7fb5b1a
	isConstructed: True
	isBuilt: True
	isInitialised: False
	hasExecuted: False
	isDestroyed: False
	constructor function name: c-Construct
	build function name: c-Build
	initialise function name: c-Initialise
	execute function name: c-Execute
	destroy function name: c-Destroy
Variable (ptr): (0x805cdd8)
	offsetCount: 1
	offsets (ptr): 0x805d0c0
		offsets[0]: 16
	dataTypes (ptr): 0x805d0d0
		dataTypes[0]: 5
	dataTypeCounts (ptr): 0x805d0e0
		dataTypeCounts[0]: 1
	dataSizes (ptr): 0x805d168
		dataSizes[0]: 8
	components (ptr): 0x805d0f0
		components[0] (ptr): (nil)
	structSize: 24
	structSizePtr (ptr): (nil)
	arrayPtr (ptr): 0x805be48
	arrayPtrPtr (ptr): 0xbfcdcde8
	arraySize: 10
	arraySizePtr (ptr): 0xbfcdcde4
	allocateSelf = False
Stg_Class (ptr): 0x805b7a0
	sizeOfSelf: 40
	_deleteSelf: Yes
	type: Variable_Register
	_delete (func ptr): 0xb7fb6561
	_print (func ptr): 0xb7fb65a1
	_copy (func ptr): 0xb7fb66ca
Setting value 123456789.000000
~~~Scalar~~~
i=0: 1 1 1 1 1 
i=1: 1 1 1 1 1 
i=2: 1 1 1 1 1 
i=3: 1 1 1 1 1 
i=4: 1 1 1 1 1 
i=5: 1 1 1 1 1 
i=6: 1 1 1 1 1 
i=7: 1 1 1 1 1 
i=8: 1 1 1 1 1 
i=9: 1 1 1 1 1 
~~~Vector~~~
a, i=0: 1 1 1 1 1 ptr test OK
b, i=0: 1 1 1 1 1 ptr test OK
c, i=0: 1 1 1 1 1 ptr test OK
a, i=1: 1 1 1 1 1 ptr test OK
b, i=1: 1 1 1 1 1 ptr test OK
c, i=1: 1 1 1 1 1 ptr test OK
a, i=2: 1 1 1 1 1 ptr test OK
b, i=2: 1 1 1 1 1 ptr test OK
c, i=2: 1 1 1 1 1 ptr test OK
a, i=3: 1 1 1 1 1 ptr test OK
b, i=3: 1 1 1 1 1 ptr test OK
c, i=3: 1 1 1 1 1 ptr test OK
a, i=4: 1 1 1 1 1 ptr test OK
b, i=4: 1 1 1 1 1 ptr test OK
c, i=4: 1 1 1 1 1 ptr test OK
a, i=5: 1 1 1 1 1 ptr test OK
b, i=5: 1 1 1 1 1 ptr test OK
c, i=5: 1 1 1 1 1 ptr test OK
a, i=6: 1 1 1 1 1 ptr test OK
b, i=6: 1 1 1 1 1 ptr test OK
c, i=6: 1 1 1 1 1 ptr test OK
a, i=7: 1 1 1 1 1 ptr test OK
b, i=7: 1 1 1 1 1 ptr test OK
c, i=7: 1 1 1 1 1 ptr test OK
a, i=8: 1 1 1 1 1 ptr test OK
b, i=8: 1 1 1 1 1 ptr test OK
c, i=8: 1 1 1 1 1 ptr test OK
a, i=9: 1 1 1 1 1 ptr test OK
b, i=9: 1 1 1 1 1 ptr test OK
c, i=9: 1 1 1 1 1 ptr test OK
~~~Vector: Sub-Variable~~~
j=0, i=0: OK
j=1, i=0: OK
j=2, i=0: OK
j=0, i=1: OK
j=1, i=1: OK
j=2, i=1: OK
j=0, i=2: OK
j=1, i=2: OK
j=2, i=2: OK
j=0, i=3: OK
j=1, i=3: OK
j=2, i=3: OK
j=0, i=4: OK
j=1, i=4: OK
j=2, i=4: OK
j=0, i=5: OK
j=1, i=5: OK
j=2, i=5: OK
j=0, i=6: OK
j=1, i=6: OK
j=2, i=6: OK
j=0, i=7: OK
j=1, i=7: OK
j=2, i=7: OK
j=0, i=8: OK
j=1, i=8: OK
j=2, i=8: OK
j=0, i=9: OK
j=1, i=9: OK
j=2, i=9: OK
Setting value 0.987654
~~~Scalar~~~
i=0: 1 1 1 1 1 
i=1: 1 1 1 1 1 
i=2: 1 1 1 1 1 
i=3: 1 1 1 1 1 
i=4: 1 1 1 1 1 
i=5: 1 1 1 1 1 
i=6: 1 1 1 1 1 
i=7: 1 1 1 1 1 
i=8: 1 1 1 1 1 
i=9: 1 1 1 1 1 
~~~Vector~~~
a, i=0: 1 1 1 1 1 ptr test OK
b, i=0: 1 1 1 1 1 ptr test OK
c, i=0: 1 1 1 1 1 ptr test OK
a, i=1: 1 1 1 1 1 ptr test OK
b, i=1: 1 1 1 1 1 ptr test OK
c, i=1: 1 1 1 1 1 ptr test OK
a, i=2: 1 1 1 1 1 ptr test OK
b, i=2: 1 1 1 1 1 ptr test OK
c, i=2: 1 1 1 1 1 ptr test OK
a, i=3: 1 1 1 1 1 ptr test OK
b, i=3: 1 1 1 1 1 ptr test OK
c, i=3: 1 1 1 1 1 ptr test OK
a, i=4: 1 1 1 1 1 ptr test OK
b, i=4: 1 1 1 1 1 ptr test OK
c, i=4: 1 1 1 1 1 ptr test OK
a, i=5: 1 1 1 1 1 ptr test OK
b, i=5: 1 1 1 1 1 ptr test OK
c, i=5: 1 1 1 1 1 ptr test OK
a, i=6: 1 1 1 1 1 ptr test OK
b, i=6: 1 1 1 1 1 ptr test OK
c, i=6: 1 1 1 1 1 ptr test OK
a, i=7: 1 1 1 1 1 ptr test OK
b, i=7: 1 1 1 1 1 ptr test OK
c, i=7: 1 1 1 1 1 ptr test OK
a, i=8: 1 1 1 1 1 ptr test OK
b, i=8: 1 1 1 1 1 ptr test OK
c, i=8: 1 1 1 1 1 ptr test OK
a, i=9: 1 1 1 1 1 ptr test OK
b, i=9: 1 1 1 1 1 ptr test OK
c, i=9: 1 1 1 1 1 ptr test OK
~~~Vector: Sub-Variable~~~
j=0, i=0: OK
j=1, i=0: OK
j=2, i=0: OK
j=0, i=1: OK
j=1, i=1: OK
j=2, i=1: OK
j=0, i=2: OK
j=1, i=2: OK
j=2, i=2: OK
j=0, i=3: OK
j=1, i=3: OK
j=2, i=3: OK
j=0, i=4: OK
j=1, i=4: OK
j=2, i=4: OK
j=0, i=5: OK
j=1, i=5: OK
j=2, i=5: OK
j=0, i=6: OK
j=1, i=6: OK
j=2, i=6: OK
j=0, i=7: OK
j=1, i=7: OK
j=2, i=7: OK
j=0, i=8: OK
j=1, i=8: OK
j=2, i=8: OK
j=0, i=9: OK
j=1, i=9: OK
j=2, i=9: OK
